{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "480bc70c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import multiprocessing\n",
    "\n",
    "from MS2ORF import MS2ORF\n",
    "\n",
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('max_colwidth', 400)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2facfa8",
   "metadata": {},
   "source": [
    "<H1>Pre-BLAST data analysis</H1>\n",
    "<H4>Integrate raw .mgf files, Metamorpheus output, and Casanovo output</H4>\n",
    "<H4>Create BLAST query file for peptide-genome search</H4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cfbb4639",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/peter/MS2ORF/MS2ORF/MS2ORF.py:94: DtypeWarning: Columns (11,20,21,22,23,24) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  temp_df1 = pd.read_csv(filename1, sep='\\t')\n",
      "/var/folders/b5/prg17fj579zd64gy82n9frm00000gn/T/ipykernel_43093/676498515.py:55: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df12 = pd.read_csv(all_dataframe_savefile1)\n",
      "/Users/peter/MS2ORF/MS2ORF/MS2ORF.py:333: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  temp_concat_sequences = pd.Series([])\n",
      "/Users/peter/MS2ORF/MS2ORF/MS2ORF.py:334: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  temp_all_concat_sequences = pd.Series([])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "780.1210589408875\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "#-------------------------\n",
    "#-----user input variables\n",
    "#-------------------------\n",
    "\n",
    "input_mgf_file_rawconverter1 = 'raw_data/'\n",
    "input_mgf_file_msconvert1 = 'raw_data/'\n",
    "metamorpheus_allPSMs_file1 = 'raw_data/AllPSMs.psmtsv'\n",
    "metamorpheus_qvalue_cutoff1 = 0.1\n",
    "casanovo_output_msconvert1 = 'raw_data/'\n",
    "casanovo_output_rawconverter1 = 'raw_data/'\n",
    "all_dataframe_savefile1 = 'processed_data/casanovo_compare1.csv'\n",
    "casanovo_PTM_list1 = ['+57.021', '+15.995', '+0.984']\n",
    "casanovo_PTM_score_threshold_for_removal1 = 90\n",
    "\n",
    "blast_query_file_name1 = 'processed_data/blastfile1.fasta'\n",
    "blast_query_file_condensed_name1 = 'processed_data/blastfile1_condensed.fasta'\n",
    "\n",
    "whole_genome_DB1 = 'raw_data/'\n",
    "transcriptome_proteome_DB1 = 'raw_data/'\n",
    "output_tBLASTn_file1 = 'processed_data/blastfile1_condensed_out.csv'\n",
    "output_BLASTp_file1 = 'processed_data/blastfile1_protdb_condensed_out.csv'\n",
    "\n",
    "#-------------------------\n",
    "#-------------------------\n",
    "\n",
    "#-----read .mgf files into dataframes\n",
    "df1 = MS2ORF.read_mgf_headers(input_mgf_file_rawconverter1, 'RC')\n",
    "df2 = MS2ORF.read_mgf_headers(input_mgf_file_msconvert1, 'MSC')\n",
    "\n",
    "#-----merge .mgf dataframes\n",
    "df3 = pd.merge(df1, df2, on='spectrum_ID', how='outer')\n",
    "df3 = df3[['scan_RC', 'spectrum_ID', 'retention_RC', 'charge_RC', 'precursor_RC', 'retention_MSC', 'charge_MSC', 'precursor_MSC']]\n",
    "df3.columns = ['scan', 'spectrum_ID', 'retention_RC', 'charge_RC', 'precursor_RC', 'retention_MSC', 'charge_MSC', 'precursor_MSC']\n",
    "\n",
    "#-----read morpheus file into dataframe\n",
    "df4 = MS2ORF.read_Morpheus(metamorpheus_allPSMs_file1, metamorpheus_qvalue_cutoff1)\n",
    "\n",
    "#-----merge .mgf and morpheus dataframes\n",
    "df5 = pd.merge(df3, df4, on='scan', how='outer').fillna(0)\n",
    "\n",
    "#-----read Casanovo output files into dataframes, then merge into single dataframe\n",
    "df6 = MS2ORF.read_Casanovo(casanovo_output_msconvert1, 'MSC')\n",
    "df7 = MS2ORF.read_Casanovo(casanovo_output_rawconverter1, 'RC')\n",
    "df8 = pd.merge(df6, df7, on='spectrum_ID', how='outer').fillna(0)\n",
    "\n",
    "#-----merge .mgf and morpheus dataframes with casanovo dataframes\n",
    "df9 = pd.merge (df5, df8, on='spectrum_ID', how='outer').fillna(0)\n",
    "\n",
    "#-----reformat dataframe for sequence comparison\n",
    "df10 = MS2ORF.reformat_Columns_For_Comparison(df9, ['denovo_seq_MSC', 'denovo_seq_RC', 'sequence'])\n",
    "\n",
    "#-----compare sequences pairwise\n",
    "df11 = MS2ORF.compare_Sequences(df10)\n",
    "\n",
    "#-----save dataframe to external file\n",
    "df11.to_csv(all_dataframe_savefile1, index=False)\n",
    "\n",
    "#-----read in dataframe csv\n",
    "df12 = pd.read_csv(all_dataframe_savefile1)\n",
    "\n",
    "#-----reformat casanovo amino acid scores (*100, then turn into integer)\n",
    "df13 = MS2ORF.reformat_aa_Scores(df12, ['aa_scores_RC', 'aa_scores_MSC'])\n",
    "\n",
    "#-----remove PTM notation and mask denovo sequences below a score threshold\n",
    "df14 = MS2ORF.mask_denovo_sequence(df13, casanovo_PTM_list1, casanovo_PTM_score_threshold_for_removal1)\n",
    "\n",
    "#-----create blast fasta file\n",
    "MS2ORF.create_BLAST_query_file(df14, ['MSC', 'RC'], blast_query_file_name1)\n",
    "\n",
    "#-----reduce size of blast file by condensing redundant peptide queries\n",
    "#-----following function produces dictionary that connects peptides to their parent spectra/table entries and also outputs a reformatted fasta for BLAST\n",
    "peptide_description_dictionary1, seqnum_peptide_dictionary1 = MS2ORF.reduce_Blast_file(blast_query_file_name1, blast_query_file_condensed_name1)\n",
    "\n",
    "#-----command for running tblastn in multithread mode using 12 threads (uncomment ! lines to enable BLAST from this notebook)\n",
    "#-----note that BLAST databases need to be built prior to running these commands [see BLAST documentation]\n",
    "#---genome\n",
    "#!tblastn -db whole_genome_DB1 -query blast_query_file_condensed_name1 -matrix IDENTITY -num_alignments 5 -outfmt 10 -num_threads 12 -mt_mode 1 -out output_tBLASTn_file1\n",
    "#---proteome/transcriptome\n",
    "#!blastp -db transcriptome_proteome_DB1 -query blast_query_file_condensed_name1 -matrix IDENTITY -num_alignments 5 -outfmt 10 -num_threads 12 -mt_mode 1 -out output_BLASTp_file1\n",
    "\n",
    "\n",
    "end = time.time()\n",
    "print(end-start)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7f9316",
   "metadata": {},
   "source": [
    "<H1>Post-BLAST data analysis</H1>\n",
    "<H4>integrate BLAST results</H4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f6b01b7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31.236429691314697\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "#-------------------------\n",
    "#-----user input variables\n",
    "#-------------------------\n",
    "\n",
    "output_tBLASTn_file1 = 'processed_data/blastfile1_condensed_out.csv'\n",
    "output_BLASTp_file1 = 'processed_data/blastfile1_protdb_condensed_out.csv'\n",
    "\n",
    "#-------------------------\n",
    "#-------------------------\n",
    "\n",
    "#-----read tBLASTn csv into dataframe\n",
    "df15 = pd.read_csv(output_tBLASTn_file1, sep=',', names=['seq_num', 'contig', 'positives_percent', 'match_length', 'mismatch_count', 'gaps', 'query_match_start', 'query_match_stop', 'target_start', 'target_stop', 'evalue', 'score'])\n",
    "\n",
    "#-----read BLASTp csv into dataframe\n",
    "df16 = pd.read_csv(output_BLASTp_file1, sep=',', names=['seq_num', 'contig', 'positives_percent', 'match_length', 'mismatch_count', 'gaps', 'query_match_start', 'query_match_stop', 'target_start', 'target_stop', 'evalue', 'score'])\n",
    "\n",
    "#-----assign peptides to seq_names in blast result dataframes\n",
    "#---df15\n",
    "pep_names_col1 = []\n",
    "for row1 in df15.itertuples():\n",
    "    pep_names_col1.append(seqnum_peptide_dictionary1[row1.seq_num])\n",
    "df15['sequence'] = pep_names_col1\n",
    "#---df16\n",
    "pep_names_col1 = []\n",
    "for row1 in df16.itertuples():\n",
    "    pep_names_col1.append(seqnum_peptide_dictionary1[row1.seq_num])\n",
    "df16['sequence'] = pep_names_col1\n",
    "\n",
    "#-----groupby peptide 'sequence' column, then select rows with smallest evalue\n",
    "#---tblastn dataframe\n",
    "df17 = df15.loc[df15.groupby('sequence').evalue.idxmin()].reset_index(drop=True)\n",
    "df17 = df17[['sequence', 'contig', 'positives_percent', 'match_length', 'mismatch_count', 'gaps', 'query_match_start', 'query_match_stop', 'target_start', 'target_stop', 'evalue', 'score']]\n",
    "#---blastp dataframe\n",
    "df18 = df16.loc[df16.groupby('sequence').evalue.idxmin()].reset_index(drop=True)\n",
    "df18 = df18[['sequence', 'contig', 'positives_percent', 'match_length', 'mismatch_count', 'gaps', 'query_match_start', 'query_match_stop', 'target_start', 'target_stop', 'evalue', 'score']]\n",
    "\n",
    "#----assign e-values to main combined dataframe peptides\n",
    "df19 = MS2ORF.associate_evalues(df14, df17, 'MSC_masked', 'genome')\n",
    "df19 = MS2ORF.associate_evalues(df19, df17, 'RC_masked', 'genome')\n",
    "df19 = MS2ORF.associate_evalues(df19, df18, 'MSC_masked', 'proteome')\n",
    "df19 = MS2ORF.associate_evalues(df19, df18, 'RC_masked', 'proteome')\n",
    "\n",
    "#-----Notes:\n",
    "#df19 contains main listing of features from database search, denovo sequencing, blast e-values, etc\n",
    "#df17 contains tblastn results condensed down to a single entry for each unique peptide (by lowest e-value)\n",
    "#df15 contains all tblastn results for comparing multiple hits\n",
    "#df18 contains blastp results condensed down to a single entry for each unique peptides (by lowest e-value)\n",
    "\n",
    "#-----output key dataframes:\n",
    "df19.to_csv('processed_data/001_combined_df.csv', sep=',', index=False)\n",
    "df17.to_csv('processed_data/001_tblastn_condensed_df.csv', sep=',', index=False)\n",
    "df15.to_csv('processed_data/001_tblastn_expanded_df.csv', sep=',', index=False)\n",
    "\n",
    "end = time.time()\n",
    "print(end-start)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "492a8d82",
   "metadata": {},
   "source": [
    "<H1>Gene sequence retrieval</H1>\n",
    "<H3>finding gene sequences in the genome from peptide clues</H3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9fed160d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/b5/prg17fj579zd64gy82n9frm00000gn/T/ipykernel_51153/2105418884.py:17: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df19 = pd.read_csv(combined_dataframe1, sep=',')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7547.325511217117\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "#-------------------------\n",
    "#-----user input variables\n",
    "#-------------------------\n",
    "\n",
    "combined_dataframe1 = 'processed_data/001_combined_df.csv'\n",
    "tblastn_condensed_dataframe1 = 'processed_data/001_tblastn_condensed_df.csv'\n",
    "tblastn_expanded_dataframe1 = 'processed_data/001_tblastn_expanded_df.csv'\n",
    "genome_fasta_file1 = 'raw_data/'\n",
    "output_denovo_ORF_fasta1 = 'processed_data/001_putative_denovoIDs.fasta'\n",
    "\n",
    "#-------------------------\n",
    "#-------------------------\n",
    "\n",
    "#-----read in csv files to dataframes\n",
    "df19 = pd.read_csv(combined_dataframe1, sep=',')\n",
    "df17 = pd.read_csv(tblastn_condensed_dataframe1, sep=',')\n",
    "df15 = pd.read_csv(tblastn_expanded_dataframe1, sep=',')\n",
    "\n",
    "#-----select rows from combined dataframe based on key parameters\n",
    "#-----look for peptides that have low evalues when BLASTed against genome and no IDs when blasted against proteome/transcriptome\n",
    "peptide_minlen = 5\n",
    "genome_minevalue = 0\n",
    "genome_maxevalue = 1\n",
    "proteome_minevalue = -100\n",
    "proteome_maxevalue = 0\n",
    "\n",
    "df20 = df19.loc[(df19['sequence'] == '0') &\n",
    "                ((df19['MSC_masked'].str.len() > peptide_minlen) & (df19['RC_masked'].str.len() > peptide_minlen)) &\n",
    "                ((df19['MSC_masked_evalues_genome'] < genome_maxevalue) & (df19['MSC_masked_evalues_genome'] > genome_minevalue)) &\n",
    "                ((df19['RC_masked_evalues_genome'] < genome_maxevalue) & (df19['RC_masked_evalues_genome'] > genome_minevalue)) &\n",
    "                ((df19['MSC_masked_evalues_proteome'] < proteome_maxevalue) & (df19['MSC_masked_evalues_proteome'] > proteome_minevalue)) &\n",
    "                ((df19['RC_masked_evalues_proteome'] < proteome_maxevalue) & (df19['RC_masked_evalues_proteome'] > proteome_minevalue)),\n",
    "                ['scan', 'spectrum_ID', 'precursor_RC', 'charge_RC', 'peptide_score_MSC', 'peptide_score_RC', 'MSC_vs_RC', 'MSC_masked', 'RC_masked', 'MSC_masked_evalues_genome', 'RC_masked_evalues_genome', 'MSC_masked_evalues_proteome', 'RC_masked_evalues_proteome']\n",
    "               ]\n",
    "\n",
    "#-----extract peptides from combined dataframe\n",
    "peptide_list1 = []\n",
    "peptide_col_names1 = ['MSC_masked', 'RC_masked']\n",
    "for item1 in peptide_col_names1:\n",
    "    peptide_list1 += list(df20[item1])\n",
    "peptide_list1 = list(set(peptide_list1))\n",
    "\n",
    "#-----pare down tblastn dataframe to include only peptides selected in above steps\n",
    "df21 = df15.loc[df15['sequence'].isin(peptide_list1)]\n",
    "\n",
    "#-----select rows with evalues < genome_maxevalue (variable from above)\n",
    "df21 = df21.loc[df21['evalue'] < genome_maxevalue].reset_index(drop=True)\n",
    "\n",
    "#-----iterate through pared down tblastn dataframe to identify source genes and their start-stop boundaries within respective contigs\n",
    "new_seq_list1 = []\n",
    "new_start_list1 = []\n",
    "new_stop_list1 = []\n",
    "\n",
    "temp_new_seq = ''\n",
    "temp_new_start = ''\n",
    "temp_new_stop = ''\n",
    "\n",
    "for row1 in df21.itertuples():\n",
    "    temp_new_seq, temp_new_start, temp_new_stop = MS2ORF.extract_Gene_Sequence(genome_fasta_file1, row1.contig, row1.target_start, row1.target_stop)\n",
    "    new_seq_list1.append(temp_new_seq)\n",
    "    new_start_list1.append(temp_new_start)\n",
    "    new_stop_list1.append(temp_new_stop)\n",
    "\n",
    "df22 = df21.copy()\n",
    "df22['gene_sequence'] = new_seq_list1\n",
    "df22['target_gene_start'] = new_start_list1\n",
    "df22['target_gene_stop'] = new_stop_list1\n",
    "\n",
    "#-----save dataframe to .csv file\n",
    "#df22.to_csv('001_isolated_genes_df.csv', index=False)\n",
    "\n",
    "#-----create denovo protein fasta\n",
    "MS2ORF.create_Denovo_Protein_Fasta(df22, 'gene_sequence', 'denovo_casanovo', output_denovo_ORF_fasta1)\n",
    "\n",
    "end = time.time()\n",
    "print(end-start)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b6b8f2b-26c0-4cb3-982e-e2a3c7105325",
   "metadata": {},
   "source": [
    "<H1>Save Output to .csv</H1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ff789c2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df22.to_csv('001_isolated_genes_df.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33fcd3e7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
